### DISCUSSION TITLE
How to make "fzf-lua live_grep" support <ctrl-r>, like "fzf-lua files"?

### DISCUSSION THREAD

**Discussion by i11010520** [LazyVim/LazyVim#4320] [Q&A] [answered]
Supposed, `nvim` was launched at `/path/to/some`.
After a while, `neotree`'s root directory was set to `/path/to/some/well`.

If I search files by stroking `<leader><space>`, the "fzf-lua files"'s `cwd` can be switched by `<ctrl-r>`.
But "fzf-lua live_grep"(`<leader>,`) was immune, keeping its `cwd` as the second path(`/path/to/some/well`).

How to make "fzf-lua live_grep" behave like "fzf-lua files"?

**Comment by dpetka2001** [comment]
It should be propagated out of the box with the latest changes in fzf-lua. Do you have fzf-lua updated? Also did you set `filesystem.bind_to_cwd = true` in neo-tree settings? Because by default it's set to `false` by LazyVim.

**Reply by i11010520** [reply]
>  Do you have fzf-lua updated? 

Yes, the latest.

> Also did you set `filesystem.bind_to_cwd = true` in neo-tree settings? Because by default it's set to `false` by LazyVim.

If the current buffer is a file outside of `neo-tree`'s root (cwd),
"fzf-lua files" and "fzf-lua live_grep" both pop up based on the path where the current buffer's file located,
however only "fzf-lua files" supports `<ctrl-r>` to switch the base path between this current path and `cwd`,
and I wanna "fzf-lua live_grep" supports too.

**Reply by dpetka2001** [reply]
It just works out of the box if you haven't changed anything. Here's a screencast installing LazyVim from scratch and enabling fzf-lua and using `<ctrl-r` in `live_grep`.

[Screencast 2024-08-19 10:19:58.webm](https://github.com/user-attachments/assets/8b3a0ab3-1641-4eb3-b2ce-eb4f7825cbce)

**Reply by i11010520** [reply]
> It just works out of the box if you haven't changed anything. Here's a screencast installing LazyVim from scratch and enabling fzf-lua and using `<ctrl-r` in `live_grep`.
> 
>  Screencast.2024-08-19.10.19.58.webm

Got it. Thx!
Actually I custom `fzf-lua` as follows:
```
return {
  "ibhagwan/fzf-lua",
  opts = function(_, opts)
    return vim.tbl_deep_extend("force", opts, {
      fzf_opts = {
        ["--layout"] = "default",
      },

      files = {
        resume = true,
        cwd_prompt = true,
      },
      grep = {
        rg_glob = true,
        glob_flag = "--iglob",
        glob_separator = "%s%-%-",
        resume = true,
        cwd_prompt = true,
      },
    })
  end,
}
```

**Reply by dpetka2001** [reply, marked as answer]
It's the flag `resume = true`. For some reason it seems to conflict with toggling cwd/root directory. Not sure why. You should ask in `fzf-lua` repo for more information if you'd like.

**Reply by i11010520** [reply]
So it is! Thx!
I would probably trace it in `fzf-lua` later.

**Reply by ibhagwan** [reply]
Ref: https://github.com/ibhagwan/fzf-lua/discussions/1412#discussioncomment-10391240

@dpetka2001, not sure how the `ctrl-r` action is configured in LazyVim, this is how I would configure it and it's working with `resume=true`, minimal setup:
> Note you'd need the latest fzf-lua commit https://github.com/ibhagwan/fzf-lua/commit/264f5d870d603d4f90ef7ec0f61fcb5987efc1d5 to call the `header` function on resume
```lua
require("fzf-lua").setup({
  actions = {
    files = {
      ["ctrl-r"] = {
        header = function(o)
          return string.format("switch to '%s'", o.cwd ~= "~" and "~" or "~/.config/nvim")
        end,
        fn = function(_, o)
          require("fzf-lua").resume({
            cwd = o.cwd ~= vim.fn.expand("$HOME") and "~" or "~/.config/nvim"
          })
        end,
      },
    }
  },
  files = { resume = true },
  grep = { resume = true },
})
```

**Reply by dpetka2001** [reply]
@ibhagwan Not sure about this, since I'm just a simple user with little programming experience, but `<ctrl-r>` is [defined here](https://github.com/LazyVim/LazyVim/blob/12818a6cb499456f4903c5d8e68af43753ebc869/lua/lazyvim/plugins/extras/editor/fzf.lua#L67-L73) and it's just supposed to switch between current working directory and LazyVim root directory (which is usually the corresponding LSP server's `root_dir`). 

This action is indeed implemented by Folke by using the `call_opts` of the current picker instance. Then, these options are passed to `LazyVim.pick` which is a custom picker that requires either a `fzf-lua` picker or a `telescope` picker depending on which Extra is enabled by the user. Since you mentioned in the other thread that by using `resume = true` the call options will get overwritten on each picker instance resume, then I guess that makes sense why OP is observing this behavior, though I have no idea why the `files` picker shows a different behavior that is not affected by `resume = true`. 

OP could also use `<leader>sR`, which is a keymap to invoke either `fzf-lua` or `telescope` picker with `resume` option. I really don't know if the default `<ctrl-r>` action should be changed according to your example, since it's not only intended for a `resume` action. It's supposed to switch between `root_dir` and `cwd` for every picker and according to the `options` that the user passes to them (I'm sure not everyone would like to use `resume = true`). But that is up to Folke to decide in the end.

**Reply by ibhagwan** [reply]
No problem @dpetka2001, there is no rush on this, it can wait for @folke.
